// neondb/dist/lib/texts.js
import { DEFAULTS } from "./utils/args.js";

//#region src/lib/texts.ts
const INTRO_ART = `


    ▟████████████▙
    ██          ██
    ██    ▗▅▖   ██       https://neon.com
    ██    ████▙ ██       ├── /docs
    ██    ██ ▜████       └── /discord
    ██    ██   ▜█▛
    ▜██████▛


`;
const messages = {
	welcome: "Let's get you set with a Postgres database on Neon.",
	nonInteractive: "`neondb --yes` for non-interactive flow with defaults.",
	generating: "Generating your database... ",
	envKeyExistsExit: "Please try again or select a different key for your connection string.",
	connectionString: (s) => `Connection string: ${s}`,
	poolerString: (s) => `Pooler string: ${s}`,
	envSuccess: (dotEnvFile, dotEnvKey) => `Saved it to ${dotEnvFile} as ${dotEnvKey} \n\n\n`,
	happyCoding: "Happy coding! 🚀",
	databaseGenerated: (url) => `Claim your Database at: ${url}`,
	questions: {
		dotEnvFilePath: `Enter the path to your environment file (default: ${DEFAULTS.dotEnvPath})`,
		dotEnvKey: `Enter the key for the database connection string (default: ${DEFAULTS.dotEnvKey})`,
		seedPath: `Enter the path to your seed (.sql) file (default: ${DEFAULTS.seed?.path || "none"})`,
		prefix: `Enter the prefix for public environment variables (default: ${DEFAULTS.envPrefix})`
	},
	info: {
		dotEnvFileNotFound: "No .env file found, creating one.",
		userCancelled: "Prompt cancelled by user.",
		defaultEnvKey: (dotEnvKey) => `using ${dotEnvKey} as the environment variable key`,
		defaultEnvFilePath: (dotEnvPath) => `using ${dotEnvPath} as the .env file`,
		defaultPrefix: (prefix) => `using ${prefix} as the prefix for public environment variables`
	},
	errors: {
		invalidEnvFile: "Invalid .env file format \n",
		envKeyExists: (dotEnvKey, dotEnvFile) => `${dotEnvKey} already exists in ${dotEnvFile}`,
		failedToParseEnvFile: "Failed to parse .env file",
		failedToWriteEnvFile: "Failed to write .env file",
		failedToGenerateDatabase: "Failed to generate database",
		failedToOpenBrowser: "Failed to open browser",
		failedToCreateClaimableDatabase: "Failed to create claimable database",
		failedToCreatePoolerString: "Failed to create pooler string",
		failedToSaveConnectionString: "Failed to save connection string",
		failedToSavePoolerString: "Failed to save pooler string",
		failedToSaveEnvFile: "Failed to save .env file"
	}
};

//#endregion
export { INTRO_ART, messages };
