// neondb/dist/cli.js
#!/usr/bin/env node
import { DEFAULTS, getArgs } from "./lib/utils/args.js";
import { INTRO_ART, messages } from "./lib/texts.js";
import { prepEnv } from "./lib/utils/fs.js";
import { instantNeon } from "./lib/instant-neon.js";
import { claim } from "./lib/utils/claim.js";
import { validateEnvKey, validateEnvPath } from "./lib/utils/validate.js";
import { intro, isCancel, log, outro, spinner, text } from "@clack/prompts";
import { cristal } from "gradient-string";

//#region src/cli.ts
async function main() {
	const { command, yes: shouldUseDefaults,...flags } = getArgs();
	if (command === "claim") {
		await claim(flags.env || DEFAULTS.dotEnvKey, flags.prefix || DEFAULTS.envPrefix);
		return;
	}
	console.log(cristal(INTRO_ART));
	const s = spinner();
	intro(messages.welcome);
	log.info(messages.nonInteractive);
	const userInput = {};
	if (shouldUseDefaults) {
		const envPath = flags.env || DEFAULTS.dotEnvPath;
		const envKey = flags.key || DEFAULTS.dotEnvKey;
		const envPrefix = flags.prefix || DEFAULTS.envPrefix;
		prepEnv(envPath, envKey);
		s.start(messages.generating);
		await instantNeon({
			dotEnvFile: envPath,
			dotEnvKey: envKey,
			referrer: "npm:neondb/cli",
			seed: flags.seed ? {
				type: "sql-script",
				path: flags.seed
			} : DEFAULTS.seed,
			envPrefix
		});
	} else {
		/**
		* Get Env file path (e.g.: .env)
		*/
		if (flags.env) {
			const isEnvPathInvalid = validateEnvPath(flags.env);
			if (isEnvPathInvalid) {
				log.error(isEnvPathInvalid.message);
				process.exit(1);
			}
			log.step(messages.info.defaultEnvFilePath(flags.env));
			userInput.dotEnvPath = flags.env;
		} else {
			userInput.dotEnvPath = await text({
				message: messages.questions.dotEnvFilePath,
				validate: validateEnvPath
			});
			if (isCancel(userInput.dotEnvPath)) {
				outro(messages.info.userCancelled);
				process.exit(1);
			}
			if (!userInput.dotEnvPath) {
				userInput.dotEnvPath = DEFAULTS.dotEnvPath;
				log.step(messages.info.defaultEnvFilePath(userInput.dotEnvPath));
			}
		}
		if (flags.key) {
			const isEnvKeyInvalid = validateEnvKey(flags.key);
			if (isEnvKeyInvalid) {
				log.error(isEnvKeyInvalid.message);
				process.exit(1);
			}
			log.step(messages.info.defaultEnvKey(flags.key));
			userInput.dotEnvKey = flags.key;
		}
		if (!userInput.dotEnvKey) {
			userInput.dotEnvKey = await text({
				message: messages.questions.dotEnvKey,
				validate: validateEnvKey
			});
			if (isCancel(userInput.dotEnvKey)) {
				outro(messages.info.userCancelled);
				process.exit(1);
			}
			if (!userInput.dotEnvKey) {
				userInput.dotEnvKey = DEFAULTS.dotEnvKey;
				log.step(messages.info.defaultEnvKey(userInput.dotEnvKey));
			}
		}
		if (!flags.seed) {
			userInput.seed = {
				type: "sql-script",
				path: await text({ message: messages.questions.seedPath })
			};
			if (!userInput.seed?.path) userInput.seed = DEFAULTS.seed;
		} else userInput.seed = {
			type: "sql-script",
			path: flags.seed
		};
		if (flags.prefix) {
			log.step(messages.info.defaultPrefix(flags.prefix));
			userInput.envPrefix = flags.prefix;
		}
		if (!userInput.envPrefix) {
			userInput.envPrefix = await text({ message: messages.questions.prefix });
			if (isCancel(userInput.envPrefix)) {
				outro(messages.info.userCancelled);
				process.exit(1);
			}
			if (!userInput.envPrefix) {
				userInput.envPrefix = DEFAULTS.envPrefix;
				log.step(messages.info.defaultPrefix(userInput.envPrefix));
			}
		}
		prepEnv(userInput.dotEnvPath, userInput.dotEnvKey);
		s.start(messages.generating);
		await instantNeon({
			dotEnvFile: userInput.dotEnvPath,
			dotEnvKey: userInput.dotEnvKey,
			referrer: "npm:neondb/cli",
			seed: userInput.seed,
			envPrefix: userInput.envPrefix
		});
	}
	s.stop("Database generated!");
	outro(messages.happyCoding);
}
await main();
var cli_default = main;

//#endregion
export { cli_default as default };
