// neondb/dist/lib/utils/fs.js
import { messages } from "../texts.js";
import { log, outro } from "@clack/prompts";
import { closeSync, existsSync, mkdirSync, openSync, readFileSync, writeSync } from "node:fs";
import { dirname } from "node:path";
import { parse } from "dotenv";

//#region src/lib/utils/fs.ts
function splitCommands(schema) {
	return schema.split(";").map((cmd) => cmd.trim()).filter(Boolean);
}
function validateSql(sql) {
	if ((sql.match(/\(/g) || []).length !== (sql.match(/\)/g) || []).length) throw new Error("SQL has unbalanced parentheses");
	return sql;
}
function getSqlCommands(path) {
	try {
		return splitCommands(validateSql(readFileSync(path, "utf8")));
	} catch (error) {
		log.error(error instanceof Error ? error.message : "Failed to read SQL file.");
		process.exit(1);
	}
}
function getDotEnvContent(dotEnvFile) {
	if (!existsSync(dotEnvFile)) {
		log.info(messages.info.dotEnvFileNotFound);
		return {};
	}
	try {
		return parse(readFileSync(dotEnvFile));
	} catch {
		throw new Error(messages.errors.failedToParseEnvFile);
	}
}
function prepEnv(dotEnvFile, dotEnvKey) {
	try {
		if (getDotEnvContent(dotEnvFile)[dotEnvKey]) {
			log.warn(messages.errors.envKeyExists(dotEnvKey, dotEnvFile));
			outro(messages.envKeyExistsExit);
			process.exit(0);
		}
		return;
	} catch (error) {
		if (error instanceof Error && error.message === messages.errors.failedToParseEnvFile) {
			console.error(error);
			log.error(messages.errors.invalidEnvFile);
			process.exit(1);
		}
	}
}
async function writeToEnv(dotEnvFile, dotEnvKey, claimExpiresAt, claimUrl, connString, poolerString, envPrefix = "PUBLIC_") {
	if (!existsSync(dirname(dotEnvFile))) mkdirSync(dirname(dotEnvFile), { recursive: true });
	const openedFile = openSync(dotEnvFile, "a");
	writeSync(openedFile, `# Claimable DB expires at: ${claimExpiresAt.toUTCString()}
# Claim it now to your account: ${claimUrl.href}
${dotEnvKey}=${connString}
${dotEnvKey}_POOLER=${poolerString}
${envPrefix}NEON_LAUNCHPAD_CLAIM_URL=${claimUrl.href}
`);
	closeSync(openedFile);
}

//#endregion
export { getDotEnvContent, getSqlCommands, prepEnv, writeToEnv };
